src/modules/ekf2/EKF/pos_vel_estimator_beacon.cpp

```diff
@@ -0,0 +1,168 @@
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "pos_vel_estimator_beacon.hpp"
+
+#include "python/ekf_derivation/generated/compute_beacon_range_fusion.h"
+#include "python/ekf_derivation/generated/predict_covariance_beacon.h"
+
+using namespace time_literals;
+
+void PosVelFromBeacon::init(matrix::Vector<float, NUM_STATES> state_init)
+{
+	if (!_initialised) {
+		_state = state_init;
+		_P.setZero();
+		_P(VX, VX) = MAX_VEL_VARIANCE;
+		_P(VY, VY) = MAX_VEL_VARIANCE;
+		_P(X, X) = MAX_POS_VARIANCE;
+		_P(Y, Y) = MAX_POS_VARIANCE;
+		_initialised = true;
+	}
+}
+void
+PosVelFromBeacon::update(uint64_t time_now, const Vector2f delta_vel, const float dt_prediction)
+{
+	if (!_initialised) {
+		return;
+	}
+
+	predictState(delta_vel, dt_prediction);
+	predictCovariance(time_now, delta_vel);
+}
+
+void PosVelFromBeacon::predictCovariance(uint64_t time_now,
+                                         const Vector2f &delta_vel)
+{// run covariance prediction at a slower rate
+	const float dt_cov_prediction = (float) (time_now - _time_last_update) * 1e-6f;
+	if (time_now - _time_last_update < FILTER_UPDATE_RATE_US || _time_last_update == 0) {
+		if (_time_last_update == 0) {
+			_time_last_update = time_now;
+		}
+
+		return;
+	}
+	_time_last_update = time_now;
+	matrix::Matrix<float, NUM_STATES, NUM_STATES> P_new;
+	const float var = sq(_accel_noise_density * dt_cov_prediction);
+	matrix::Vector2f delta_vel_var(var, var);
+	sym::PredictCovarianceBeacon(_state, _P, delta_vel, delta_vel_var, dt_cov_prediction, &P_new);
+	_P = P_new;
+}
+
+void PosVelFromBeacon::predictState(const Vector2f &delta_vel, const float dt_prediction)
+{// predict state
+	_state(VX, 0) += delta_vel(0);
+	_state(VY, 0) += delta_vel(1);
+	_state(X, 0) += _state(VX, 0) * dt_prediction;
+	_state(Y, 0) += _state(VY, 0) * dt_prediction;
+}
+
+void PosVelFromBeacon::fuseRange(const beaconRangeSample &range_sample, const float vehicle_pos_z)
+{
+	if (!_initialised) {
+		return;
+	}
+
+	matrix::Matrix<float, NUM_STATES, 1> K;
+	matrix::Matrix<float, 1, NUM_STATES> H;
```

Comment:
Those can be vectors. Just return the transpose of H in the function used by symforce